### PR TITLE
FEA-2771 Add ActionV2 w/o null safety

### PR DESCRIPTION
**Pull request by sorenthompson-wk**
## Motivation
Adding APIs in major versions introduces risk, since consumers with opened ranges could accidentally consume new APIs. So instead of exposing Action2 in w_flux 3.0, we will backpatch Action2 to a w_flux 2.x minor, so that it won't be an API addition in w_flux 3.0. This option enables us to start work on the Action migration before the w_flux major release has finished rolling out

## Changes
Add Action2 and appropriate test coverage without null safety work.

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Architecture member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/w_flux/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/w_flux/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: test/action_test.dart [hunk 1]
```diff
@@ -37,14 +37,13 @@ void main() {
...
-      });
+      Action<String> _action = Action<String>()
+        ..listen((String payload) {
+          expect(payload, equals(null));
+          c.complete();
+        });
 
-      _action();
+      await _action();
       return c.future;
```
Comment (evanweible-wf) [inline test/action_test.dart:47]: #nit Unrelated to your changes, but these tests should probably be updated to use `expectAsync1` instead of a `Completer`. Something like:

```dart
final _action = Action<String>()
  ..listen(expectAsync1((payload) {
    expect(payload, isNull);
  });
  await _action();
```

File: test/action_test.dart [hunk 2]
```diff
@@ -118,21 +117,20 @@ void main() {
         var action = Action();
         var asyncListenerCompleted = false;
         action.listen((_) async {
-          await Future.delayed(Duration(milliseconds: 100), () {
+          await Future.delayed(const Duration(milliseconds: 100), () {
```
Comment (evanweible-wf) [inline test/action_test.dart:120]: I thought the `const` would be optional at this point, is it not?

File: test/action_test.dart [hunk 3]
```diff
@@ -118,21 +117,20 @@ void main() {
...
             asyncListenerCompleted = true;
           });
         });
 
-        var future = action();
+        Future<dynamic> future = action();
```
Comment (evanweible-wf) [inline test/action_test.dart:125]: Was this type change needed?

File: test/action_test.dart [hunk 4]
```diff
@@ -201,16 +199,231 @@ void main() {
...
+    test('should dispatch by default when called', () async {
+      Completer c = Completer();
+      action.listen((payload) {
+        expect(payload, equals('990 guerrero'));
+        c.complete();
+      });
+
+      await action('990 guerrero');
+      return c.future;
+    });
```
Comment (evanweible-wf) [inline test/action_test.dart:258]: Hmm.. I know this is copied from the test suite above, but in both cases this test seems identical to the one before it.

**Review by evanweible-wf — COMMENTED** [review]
Just some nits, otherwise LGTM. I'm also still working on a spike with Action0 and Action1 (for arity 0 and arity 1 since we talked about that alternative - we can compare them and see what we think.

File: test/action_test.dart (hunk 3 above)
Comment (sorenthompson-wk) [reply test/action_test.dart:125]: It was suggested by the dart linter: 
```
Use 'const' with the constructor to improve performance.
Try adding the 'const' keyword to the constructor invocation.
dartprefer_const_constructors
```

File: test/action_test.dart (hunk 2 above)
Comment (sorenthompson-wk) [reply test/action_test.dart:120]: It was suggested by the dart linter: 
```
Use 'const' with the constructor to improve performance.
Try adding the 'const' keyword to the constructor invocation.
dartprefer_const_constructors
```

File: test/action_test.dart (hunk 3 above)
Comment (sorenthompson-wk) [reply test/action_test.dart:125]: Nope! I changed it back.

File: lib/src/action.dart [hunk 5]
```diff
@@ -45,15 +56,15 @@ import 'package:w_flux/src/constants.dart' show v3Deprecation;
 /// when a consumer needs to check state changes immediately after invoking an
 /// action.
 ///
-class Action<T> extends Object with Disposable implements Function {
+class ActionV2<T> extends Object with Disposable implements Function {
```
Comment (todbachman-wf) [inline lib/src/action.dart:59]: #supernit. I think we could simplify this without affecting the public API.

```suggestion
class ActionV2<T> extends Disposable {
```

File: test/action_test.dart [hunk 6]
```diff
@@ -26,62 +26,48 @@ void main() {
 
     setUp(() {
       action = Action<String>();
+      addTearDown(() async {
+        await action.dispose();
+      });
```
Comment (todbachman-wf) [inline test/action_test.dart:31]: #supernit. This could be reduced to:

```suggestion
      addTearDown(action.dispose);
```

File: test/action_test.dart [hunk 7]
```diff
@@ -26,62 +26,48 @@ void main() {
...
     });
 
     test('should only be equivalent to itself', () {
       Action _action = Action();
-      Action _action2 = Action();
+      Action _ActionV2 = Action();
```
Comment (todbachman-wf) [inline test/action_test.dart:36]: It looks like this might have been a find/replace issue?

```suggestion
      Action _action2 = Action();
```

Also, I know this isn't your fault, but it feels like the underscores on these local variables are unnecessary. Might be nice to remove them if you need to make a change to this test.

File: test/action_test.dart [hunk 8]
```diff
@@ -201,16 +186,188 @@ void main() {
...
+      addTearDown(() async {
+        await action.dispose();
+      });
+    });
+
+    test('should only be equivalent to itself', () {
+      ActionV2 _action = ActionV2();
+      ActionV2 _ActionV2 = ActionV2();
+      expect(_action == _action, isTrue);
+      expect(_action == _ActionV2, isFalse);
```
Comment (todbachman-wf) [inline test/action_test.dart:225]: This was probably a find/replace error too.

```suggestion
      ActionV2 action = ActionV2();
      ActionV2 action2 = ActionV2();
      expect(action == action, isTrue);
      expect(action == action2, isFalse);
```

**Review by todbachman-wf — COMMENTED** [review]
I've just got some nits around the tests.

File: test/action_test.dart (hunk 7 above)
Comment (sorenthompson-wk) [reply test/action_test.dart:36]: oops! yes find and replace

**Review by todbachman-wf — APPROVED** [review]

File: lib/src/action.dart [hunk 9]
```diff
@@ -20,11 +20,22 @@ import 'package:w_common/disposable.dart';
 
 import 'package:w_flux/src/constants.dart' show v3Deprecation;
 
+/// Like [ActionV2], but payloads cannot be made non-nullable since the argument
+/// to [call] is optional.
+@Deprecated('Use ActionV2 instead, which supports non-nullable payloads.')
+class Action<T> extends ActionV2<T> {
```
Comment (evanweible-wf) [inline lib/src/action.dart:26]: Now that ActionV2 is the common base class, I think we will also need to update the typing in the `store.dart` file for any parameter that currently expects type `Action` to expect type `ActionV2` instead - that way they will be able to accept either an `Action` or an `ActionV2` as people migrate.

**Review by todbachman-wf — APPROVED** [review]

File: lib/src/store.dart [hunk 10]
```diff
@@ -153,7 +153,7 @@ class Store extends Stream<Store> with Disposable {
   /// called until that future has resolved.
   ///
   /// If the `Store` has been disposed, this method throws a [StateError].
-  void triggerOnActionV2<T>(Action<T> action,
+  void triggerOnActionV2<T>(ActionV2<T> action,
```
Comment (evanweible-wf) [inline lib/src/store.dart:156]: 😆 this method name now works on two different levels

**Review by evanweible-wf — APPROVED** [review]

**Review by todbachman-wf — APPROVED** [review]

**Comment by sorenthompson-wk** [comment]
QA +1 
- I saw the deprecation message for Action in my IDE.
- I ran the example app of w_outline with this branch of w_flux and the example app performed as expected.

**Comment by sorenthompson-wk** [comment]
@Workiva/release-management-p ready to merge

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM
